src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala

```diff
@@ -90,138 +92,151 @@ class FastSync(
+        val mptNodes = handlerState.requestedMptNodes
+        val nonMptNodes = handlerState.requestedNonMptNodes
+        val requestedHashes = mptNodes.getOrElse(sender(), Nil) ++ nonMptNodes.getOrElse(sender(), Nil)
+        context unwatch sender()
+        val newState = handlerState
+          .withRequestedMptNodes(mptNodes - sender())
+          .withRequestedNonMptNodes(nonMptNodes - sender())
+          .withAssignedHandlers(handlerState.assignedHandlers - sender())
+        handleNodeData(peer, requestedHashes, nodeData, newState)
+    }
+
+    def waitingForTargetBlockUpdate(processState: FinalBlockProcessingResult, handlerState: HandlerState): Receive =
+      handleCommonMessages orElse {
       case FastSyncTargetBlockSelector.Result(targetBlockHeader) =>
-        log.info(s"new target block with number ${targetBlockHeader.number} received")
-        if (targetBlockHeader.number >= syncState.targetBlock.number) {
-          updateTargetSyncState(processState, targetBlockHeader)
-          syncState = syncState.copy(updatingTargetBlock = false)
-          context become this.receive
-          processSyncing()
+        log.info(s"New target block with number ${targetBlockHeader.number} received")
+        val currentSyncState = handlerState.syncState
+        if (targetBlockHeader.number >= currentSyncState.targetBlock.number) {
+          val handlerStateBeforeUpdate = handlerState.withSyncState(currentSyncState.copy(updatingTargetBlock = false))
+          val syncStateAfterUpdate = updateTargetSyncState(processState, targetBlockHeader, handlerStateBeforeUpdate)
+
+          val newHandlerState = handlerStateBeforeUpdate.withSyncState(syncStateAfterUpdate)
+          context become receive(newHandlerState)
+          self ! ProcessSyncing
         } else {
-          syncState = syncState.copy(targetBlockUpdateFailures = syncState.targetBlockUpdateFailures + 1)
+          val newSyncState = currentSyncState.copy(targetBlockUpdateFailures = currentSyncState.targetBlockUpdateFailures + 1)
+          context become receive(handlerState.withSyncState(newSyncState))
           scheduler.scheduleOnce(syncRetryInterval, self, UpdateTargetBlock(processState))
         }
 
-      case PersistSyncState => persistSyncState()
+      case PersistSyncState => persistSyncState(handlerState)
 
-      case UpdateTargetBlock(state) => updateTargetBlock(state)
+      case UpdateTargetBlock(state) => updateTargetBlock(state, handlerState)
     }
 
-    private def updateTargetBlock(state: FinalBlockProcessingResult): Unit = {
-      syncState = syncState.copy(updatingTargetBlock = true)
-      if (syncState.targetBlockUpdateFailures <= syncConfig.maximumTargetUpdateFailures) {
-        if (assignedHandlers.nonEmpty) {
-          log.info(s"Still waiting for some responses, rescheduling target block update")
+    private def updateTargetBlock(state: FinalBlockProcessingResult, handlerState: HandlerState): Unit = {
```

Comment:
The same here. In my opinion, `updateTargetBlock` is a responsibility of `handlerState`. By the way, I suggest trying to avoid methods returning `Unit` and switching actor contexts. Is very hard to reasoning about the program flow. Maybe it is possible to make this method returning state and lift the context switching to the higher level.